working/0546-patterns/records-feature-specification.md

```diff
@@ -283,6 +283,41 @@ fields are) and collection literals.
 
 **TODO: Specify this more precisely.**
 
+### Constants
+
+A record expression is a compile-time constant expression
+if all its record field expressions are compile-time constant expressions. 
+
+_This is true whether the expression occurs in a constant context or not,
+which means that a record expression can be used directly as a parameter default value 
+if its record field expressions are constant expressions.
+Example: `f({(int, int) x = (1, 2)}) => ...`._
+
+A record expression is a potentially constant expression 
+if all its record field expressions are potentially constant or constant expressions.
+
+_This means that a record expression can be used in the initializer list
+of a constant non-redirecting generative constructor._
+
+Constant *object* instantiations create deeply immutable and canonicalied objects.
+Records are always unmodifiable, and if their field values are deeply immutable,
+like constants values, the records are also deeply immutable.
+It's meaningless to consider whether record constants are canonicalized,
+since records do not have a persistent identity.
```

Comment:
The purpose of having explicitly constant records and specifying canonicalization would be that (1) this clarifies the intent, and makes it an error if some components of a `const` record literal are not constant expressions, and (2) this serves as a request for physical canonicalization. Granted, if there is a location in code where this kind of record is unboxed then we would _not_ promise to find the canonical object if and when it is re-boxed. Still, I'd expect the explicit request for canonicalization to be helpful, in order to make `identical` work as a fast path for testing equality.

Reply:
If you plan to depend on canonicalization, you're setting yourself up for failure.

I'd expect a significant percentage of record-typed instance variables to be inlined in the object structure. You lose canonicalization for those immediately at creation time. 

Asking for canonicalization is asking for something you should not expect getting, which is why I actually prefer *not* giving you the option.

Reply:
> If you plan to depend on canonicalization, you're setting yourself up for failure.

That is true!

But it still clarifies the intent and allows for compile-time checking that we have a constant.

Of course, a top level `const freshVariableName = ...` would do that, too. So the question is just whether or not we want this syntactic sugar.

Reply:
Yeah, I think canonicalization is a non-starter for records. The language has no formal notion of what it means to assign a value or bind it to a variable. So even if you *have* a canonicalized constant record, there's no guarantee that when you *mention* it that doing so doesn't copy it to a thing with a new identity. Even binding it to the parameter to the `identical()` function itself could cause the canonicalized record to be copied to a new stack-allocated storage and lose its original identity.